In [1]:
import warnings
warnings.filterwarnings(action='ignore')
import numpy as np
import pandas as pd

pandas의 데이터프레임과 시리즈

데이터프레임과 시리즈는 파이썬의 리스트나 딕셔너리와 비슷하지만 데이터를 다루는데 더 특화되어 있다.  
리스트나 딕셔너리는 많은 양의 데이터를 저장하거나 조작할 수 있는 제공되는 메소드가 많치 않은 반면에 pandas의 데이터프레임과 시리즈는 많은 양의 데이터를 처리할 수 있을 뿐만아니라 스프레드시트 계열(엘셀 등)의 프로그램을 사용하는 것 처럼 행과 열 단위로 데이터를 조작할 수 있는 다양한 메소드와 속성을 제공한다.

시리즈 만들기

시리즈를 만들기 위해서는 리스트 또는 튜플을 Series() 클래스의 인수로 전달하면 된다.

In [2]:
s = pd.Series(['banana', 40])
print(type(s))
print(s)

<class 'pandas.Series'>
0    banana
1        40
dtype: object


In [3]:
s = pd.Series(('홍길동', '임꺽정'))
print(type(s))
print(s)
print(s.loc[0])
print(s.iloc[1])

<class 'pandas.Series'>
0    홍길동
1    임꺽정
dtype: str
홍길동
임꺽정


시리즈를 생성할때 문자열을 인덱스로 지정할 수 있다.  
문자열을 인덱스로 지정하려면 index 속성으로 인덱스로 사용할 문자열을 리스트에 담아서 전달하면 된다.

In [4]:
s = pd.Series(('홍길동', '임꺽정'), index=['name1', 'name2'])
print(s)
print(s.index)
print(s.loc['name1'])
print(s.iloc[0])

name1    홍길동
name2    임꺽정
dtype: str
Index(['name1', 'name2'], dtype='str')
홍길동
홍길동


데이터프레임 만들기

데이터프레임을 만들기 위해서는 딕셔너리 DataFrame() 클래스의 인수로 전달하면 된다.  
딕셔너리의 key가 데이터프레임의 열 이름이 되고 딕셔너리의 value(반드시 리스트 형태)는 데이터가 된다.

In [5]:
data = {
    'name': ['홍길동', '임꺽정'],
    'job': ['도둑넘', '도적놈'],
    'born': ['1569-04-20', '1875-06-10'],
    'died': ['1620-08-25', '1950-01-15'],
    'age': [51, 75]
}
print(type(data))
print(data)
df = pd.DataFrame(data)
print(type(df))
df

<class 'dict'>
{'name': ['홍길동', '임꺽정'], 'job': ['도둑넘', '도적놈'], 'born': ['1569-04-20', '1875-06-10'], 'died': ['1620-08-25', '1950-01-15'], 'age': [51, 75]}
<class 'pandas.DataFrame'>


,name,job,born,died,age
0,홍길동,도둑넘,1569-04-20,1620-08-25,51
1,임꺽정,도적놈,1875-06-10,1950-01-15,75


In [6]:
data = {
    'job': ['도둑넘', '도적놈'],
    'born': ['1569-04-20', '1875-06-10'],
    'died': ['1620-08-25', '1950-01-15'],
    'age': [51, 75]
}
# 데이터프레임을 만들때 문자열을 인덱스로 지정할 수 있다.
# 문자열을 인덱스로 지정하려면 index 속성으로 인덱스로 사용할 문자열을 리스트에 담아서 전달하면 된다.
# columns 속성으로 데이터프레임으로 만들려는 딕셔너리의 key 중에서 데이터프레임으로 만들 key를 지정할 수 있다.
df = pd.DataFrame(data, index=['홍길동', '임꺽정'], columns=['job', 'age'])
df

,job,age
홍길동,도둑넘,51
임꺽정,도적놈,75


시리즈 다루기

In [7]:
data = {
    'job': ['도둑넘', '도적놈'],
    'born': ['1569-04-20', '1875-06-10'],
    'died': ['1620-08-25', '1950-01-15'],
    'age': [51, 75]
}
df = pd.DataFrame(data, index=['홍길동', '임꺽정'])
df

,job,born,died,age
홍길동,도둑넘,1569-04-20,1620-08-25,51
임꺽정,도적놈,1875-06-10,1950-01-15,75


In [8]:
# 데이터프레임에서 시리즈를 선택하려면 loc 속성에 인덱스(열 또는 행 이름)를 전달하거나 iloc 속성에 열 또는 행의 일련번호를 전달한다.
row = df.iloc[0]
print(type(row))
print(row)
print('=' * 100)
row = df.loc['홍길동']
print(type(row))
print(row)

<class 'pandas.Series'>
job            도둑넘
born    1569-04-20
died    1620-08-25
age             51
Name: 홍길동, dtype: object
<class 'pandas.Series'>
job            도둑넘
born    1569-04-20
died    1620-08-25
age             51
Name: 홍길동, dtype: object


In [9]:
col = df.iloc[:, 0]
print(type(col))
print(col)
print('=' * 100)
col = df.loc[:, 'job']
print(type(col))
print(col)

<class 'pandas.Series'>
홍길동    도둑넘
임꺽정    도적놈
Name: job, dtype: str
<class 'pandas.Series'>
홍길동    도둑넘
임꺽정    도적놈
Name: job, dtype: str


index, values 속성과 keys() 메소드

In [10]:
print(row)
print(row.index) # 시리즈의 인덱스를 얻어온다.
print(row.keys()) # keys() 메소드도 index 속성과 같이 시리즈의 인덱스를 얻어온다.
print(row.values) # 시리즈의 데이터를 얻어온다.
print(type(row.values))

job            도둑넘
born    1569-04-20
died    1620-08-25
age             51
Name: 홍길동, dtype: object
Index(['job', 'born', 'died', 'age'], dtype='str')
Index(['job', 'born', 'died', 'age'], dtype='str')
['도둑넘' '1569-04-20' '1620-08-25' np.int64(51)]
<class 'numpy.ndarray'>


In [11]:
print(col)
print(col.index)
print(col.keys())
print(col.values)
print(type(col.values))

홍길동    도둑넘
임꺽정    도적놈
Name: job, dtype: str
Index(['홍길동', '임꺽정'], dtype='str')
Index(['홍길동', '임꺽정'], dtype='str')
<StringArray>
['도둑넘', '도적놈']
Length: 2, dtype: str
<class 'pandas.arrays.StringArray'>


시리즈 통계 메소드

In [12]:
ages = df.age
print(type(ages))
print(ages)

print(f'개수: {ages.count()}')
print(f'합계: {ages.sum()}')
print(f'평균: {ages.mean()}')
print(f'최대값: {ages.max()}')
print(f'최소값: {ages.min()}')
print(f'분산: {ages.var()}')
print(f'표준편차: {ages.std()}')

<class 'pandas.Series'>
홍길동    51
임꺽정    75
Name: age, dtype: int64
개수: 2
합계: 126
평균: 63.0
최대값: 75
최소값: 51
분산: 288.0
표준편차: 16.97056274847714


In [13]:
# scientists_df = pd.read_csv('./data/scientists.csv', sep=',')
scientists_df = pd.read_csv('./data/scientists.csv') # sep 옵션의 기본값은 ','이다.
scientists_df

,Name,Born,Died,Age,Occupation
0,Rosaline Franklin,1920-07-25,1958-04-16,37,Chemist
1,William Gosset,1876-06-13,1937-10-16,61,Statistician
2,Florence Nightingale,1820-05-12,1910-08-13,90,Nurse
3,Marie Curie,1867-11-07,1934-07-04,66,Chemist
4,Rachel Carson,1907-05-27,1964-04-14,56,Biologist
5,John Snow,1813-03-15,1858-06-16,45,Physician
6,Alan Turing,1912-06-23,1954-06-07,41,Computer Scientist
7,Johann Gauss,1777-04-30,1855-02-23,77,Mathematician


In [14]:
ages = scientists_df.Age
print(type(ages))
print(ages)

print(f'개수: {ages.count()}')
print(f'합계: {ages.sum()}')
print(f'평균: {ages.mean()}')
print(f'최대값: {ages.max()}')
print(f'최소값: {ages.min()}')
print(f'분산: {ages.var()}')
print(f'표준편차: {ages.std()}')

<class 'pandas.Series'>
0    37
1    61
2    90
3    66
4    56
5    45
6    41
7    77
Name: Age, dtype: int64
개수: 8
합계: 473
평균: 59.125
최대값: 90
최소값: 37
분산: 335.8392857142857
표준편차: 18.325918413937288


브로드캐스팅

시리즈나 데이터프레임에서 여러 데이터에 대해서 한 번에 연산하는 것을 브로드캐스팅이라 한다. 다대일 연산  
여러개의 값을 가지는 데이터 벡터라하고 단순 크기를 나타내는 데이터 1개를 스칼라라 부른다.

scientists_df 데이터프레임의 Age 시리즈에서 평균 나이(59.125)보다 나이가 많은 데이터만 추출한다.

In [15]:
bool_values = ages > ages.mean() # 브로드캐스팅
print(bool_values)

0    False
1     True
2     True
3     True
4    False
5    False
6    False
7     True
Name: Age, dtype: bool


In [16]:
print(ages[bool_values]) # 불린 인덱싱

1    61
2    90
3    66
7    77
Name: Age, dtype: int64


In [17]:
print(ages[[False, True, True, True, False, False, False, True]])

1    61
2    90
3    66
7    77
Name: Age, dtype: int64


In [18]:
print(ages[ages > ages.mean()]) # 브로드캐스팅을 이용한 불린 인덱싱

1    61
2    90
3    66
7    77
Name: Age, dtype: int64


In [19]:
for age in ages:
    print('{:4d}'.format(age), end=' ')

  37   61   90   66   56   45   41   77 

In [20]:
# 같은 길이의 벡터로 연산을 하면 결과값으로 같은 길이의 벡터가 출력된다.
for age in ages + ages:
    print('{:4d}'.format(age), end=' ')
print()
for age in ages * ages:
    print('{:4d}'.format(age), end=' ')

  74  122  180  132  112   90   82  154 
1369 3721 8100 4356 3136 2025 1681 5929 

In [21]:
# 벡터에 스칼라 값을 연산하면 각각의 벡터값에 스칼라 값이 연산된 결과가 출력된다.
for age in ages + 100:
    print('{:4d}'.format(age), end=' ')
print()
for age in ages * 2:
    print('{:4d}'.format(age), end=' ')

 137  161  190  166  156  145  141  177 
  74  122  180  132  112   90   82  154 

In [22]:
# 길이가 서로 다른 벡터를 연산하면 같은 인덱스를 가지는 값만 계산된다.
print(pd.Series([1, 100]))
# pd.Series([1, 100])의 0, 1번째 인덱스만 계산되고 나머지는 계산할 수 없으므로 누락값(결측치, NaN, nan, NA, na)으로 처리된다.
print(ages + pd.Series([1, 100]))

0      1
1    100
dtype: int64
0     38.0
1    161.0
2      NaN
3      NaN
4      NaN
5      NaN
6      NaN
7      NaN
dtype: float64


시리즈 정렬

sort_index() 메소드는 인덱스를 정렬하고 sort_values() 메소드는 데이터를 정렬한다.  
sort_index() 메소드와 sort_values() 메소드는 별도의 설정을 하지않으면 오름차순으로 정렬되지만 내림차순으로 정렬하려면 ascending 속성값을 False로 지정한다.

In [23]:
print(ages)
print(ages.sort_index()) # 오름차순 정렬, ages.sort_index(ascending=True)
print(ages.sort_index(ascending=False)) # 내림차순 정렬

0    37
1    61
2    90
3    66
4    56
5    45
6    41
7    77
Name: Age, dtype: int64
0    37
1    61
2    90
3    66
4    56
5    45
6    41
7    77
Name: Age, dtype: int64
7    77
6    41
5    45
4    56
3    66
2    90
1    61
0    37
Name: Age, dtype: int64


In [24]:
print(ages)
print(ages.sort_values()) # 오름차순 정렬, ages.sort_values(ascending=True)
print(ages.sort_values(ascending=False)) # 내림차순 정렬

0    37
1    61
2    90
3    66
4    56
5    45
6    41
7    77
Name: Age, dtype: int64
0    37
6    41
5    45
4    56
1    61
3    66
7    77
2    90
Name: Age, dtype: int64
2    90
7    77
3    66
1    61
4    56
5    45
6    41
0    37
Name: Age, dtype: int64


In [25]:
# 벡터와 벡터의 연산은 같은 인덱스를 가지는 값끼리 수행된다.
for age in ages + ages.sort_index(ascending=False):
    print('{:4d}'.format(age), end=' ')

  74  122  180  132  112   90   82  154 

데이터프레임도 시리즈와 마찬가지로 브로드캐스팅과 불린 인덱싱이 가능하다.

In [26]:
scientists_df

,Name,Born,Died,Age,Occupation
0,Rosaline Franklin,1920-07-25,1958-04-16,37,Chemist
1,William Gosset,1876-06-13,1937-10-16,61,Statistician
2,Florence Nightingale,1820-05-12,1910-08-13,90,Nurse
3,Marie Curie,1867-11-07,1934-07-04,66,Chemist
4,Rachel Carson,1907-05-27,1964-04-14,56,Biologist
5,John Snow,1813-03-15,1858-06-16,45,Physician
6,Alan Turing,1912-06-23,1954-06-07,41,Computer Scientist
7,Johann Gauss,1777-04-30,1855-02-23,77,Mathematician


In [27]:
# ages[ages > ages.mean()]
scientists_df[scientists_df.Age > scientists_df.Age.mean()]

,Name,Born,Died,Age,Occupation
1,William Gosset,1876-06-13,1937-10-16,61,Statistician
2,Florence Nightingale,1820-05-12,1910-08-13,90,Nurse
3,Marie Curie,1867-11-07,1934-07-04,66,Chemist
7,Johann Gauss,1777-04-30,1855-02-23,77,Mathematician


In [28]:
# 시리즈에 스칼라 연산을 적용할때 모든 요소에 스칼라 연산이 실행되는 것 처럼 데이터프레임도 스칼라 연산이 적용된다.
# 데이터프레임에 2를 곱하면 숫자 데이터는 2를 곱한 숫자가 되고 문자열 데이터는 2회 반복된다.
scientists_df * 2

,Name,Born,Died,Age,Occupation
0,Rosaline FranklinRosaline Franklin,1920-07-251920-07-25,1958-04-161958-04-16,74,ChemistChemist
1,William GossetWilliam Gosset,1876-06-131876-06-13,1937-10-161937-10-16,122,StatisticianStatistician
2,Florence NightingaleFlorence Nightingale,1820-05-121820-05-12,1910-08-131910-08-13,180,NurseNurse
3,Marie CurieMarie Curie,1867-11-071867-11-07,1934-07-041934-07-04,132,ChemistChemist
4,Rachel CarsonRachel Carson,1907-05-271907-05-27,1964-04-141964-04-14,112,BiologistBiologist
5,John SnowJohn Snow,1813-03-151813-03-15,1858-06-161858-06-16,90,PhysicianPhysician
6,Alan TuringAlan Turing,1912-06-231912-06-23,1954-06-071954-06-07,82,Computer ScientistComputer Scientist
7,Johann GaussJohann Gauss,1777-04-301777-04-30,1855-02-231855-02-23,154,MathematicianMathematician


데이터프레임 열의 자료형 바꾸기

In [29]:
print(scientists_df.dtypes)

Name            str
Born            str
Died            str
Age           int64
Occupation      str
dtype: object


날짜를 문자열 형태로 저장한 데이터는 날짜 및 시간에 관련된 작업을 할 수 있도록 데이터 타입을 변경해야 한다.  
to_datetime() 메소드는 str(문자열) 타입의 데이터를 datetime(날짜 및 시간) 타입으로 변경한다.

In [30]:
born_datetime = pd.to_datetime(scientists_df.Born)
print(born_datetime)
died_datetime = pd.to_datetime(scientists_df.Died)
print(died_datetime)

0   1920-07-25
1   1876-06-13
2   1820-05-12
3   1867-11-07
4   1907-05-27
5   1813-03-15
6   1912-06-23
7   1777-04-30
Name: Born, dtype: datetime64[us]
0   1958-04-16
1   1937-10-16
2   1910-08-13
3   1934-07-04
4   1964-04-14
5   1858-06-16
6   1954-06-07
7   1855-02-23
Name: Died, dtype: datetime64[us]


In [31]:
scientists_df.Born = pd.to_datetime(scientists_df.Born)
scientists_df.Died = pd.to_datetime(scientists_df.Died)

In [32]:
scientists_df.dtypes

Name                     str
Born          datetime64[us]
Died          datetime64[us]
Age                    int64
Occupation               str
dtype: object

파생 변수(새로운 열, 계산에 의한 변수) 추가하기 => 데이터프레임의 한 열을 변수라고도 부른다.

In [33]:
# str 타입을 datetime 타입으로 변경한 born_datetime, died_datetime를 데이터프레임에 파생 변수로 추가한다.
print(scientists_df.shape)
# 데이터프레임에 존재하는 열에 데이터를 넣으면 기존 데이터가 수정되고 존재하지 않는 열에 데이터를 넣으면 데이터가 추가된다.
# 데이터프레임에 파생 변수 추가는 데이터프레임이름['파생변수이름'] = 데이터 형태만 가능하다.
scientists_df['born_dt'] = born_datetime
# scientists_df.get('born_dt') = born_datetime # 사용 불가능
# scientists_df.born_dt = born_datetime # 사용 불가능
print(scientists_df.shape)
scientists_df['died_dt'] = died_datetime
print(scientists_df.shape)

(8, 5)
(8, 6)
(8, 7)


In [34]:
scientists_df['age_days'] = scientists_df.died_dt - scientists_df.born_dt
print(scientists_df.shape)
scientists_df

(8, 8)


,Name,Born,Died,Age,Occupation,born_dt,died_dt,age_days
0,Rosaline Franklin,1920-07-25,1958-04-16,37,Chemist,1920-07-25,1958-04-16,13779 days
1,William Gosset,1876-06-13,1937-10-16,61,Statistician,1876-06-13,1937-10-16,22404 days
2,Florence Nightingale,1820-05-12,1910-08-13,90,Nurse,1820-05-12,1910-08-13,32964 days
3,Marie Curie,1867-11-07,1934-07-04,66,Chemist,1867-11-07,1934-07-04,24345 days
4,Rachel Carson,1907-05-27,1964-04-14,56,Biologist,1907-05-27,1964-04-14,20777 days
5,John Snow,1813-03-15,1858-06-16,45,Physician,1813-03-15,1858-06-16,16529 days
6,Alan Turing,1912-06-23,1954-06-07,41,Computer Scientist,1912-06-23,1954-06-07,15324 days
7,Johann Gauss,1777-04-30,1855-02-23,77,Mathematician,1777-04-30,1855-02-23,28422 days


In [35]:
print(scientists_df.dtypes)

Name                      str
Born           datetime64[us]
Died           datetime64[us]
Age                     int64
Occupation                str
born_dt        datetime64[us]
died_dt        datetime64[us]
age_days      timedelta64[us]
dtype: object


데이터프레임의 열 삭제하기

drop() 메소드는 데이터베이스의 열을 삭제한다.  
drop() 메소드는 axis=0이 기본값이므로 열 방향으로 작업을 하기 때문에 행을 삭제하려 하기 때문에 에러가 발생되므로 axis=1로 설정해야 한다.

In [36]:
# scientists_drop = scientists_df.drop('Born') # scientists_df.drop('Born', axis=0), 에러
# scientists_drop = scientists_df.drop('Born', axis=1) # 열 1개 삭제
scientists_drop = scientists_df.drop(['Born', 'Died'], axis=1)

In [37]:
scientists_drop

,Name,Age,Occupation,born_dt,died_dt,age_days
0,Rosaline Franklin,37,Chemist,1920-07-25,1958-04-16,13779 days
1,William Gosset,61,Statistician,1876-06-13,1937-10-16,22404 days
2,Florence Nightingale,90,Nurse,1820-05-12,1910-08-13,32964 days
3,Marie Curie,66,Chemist,1867-11-07,1934-07-04,24345 days
4,Rachel Carson,56,Biologist,1907-05-27,1964-04-14,20777 days
5,John Snow,45,Physician,1813-03-15,1858-06-16,16529 days
6,Alan Turing,41,Computer Scientist,1912-06-23,1954-06-07,15324 days
7,Johann Gauss,77,Mathematician,1777-04-30,1855-02-23,28422 days


데이터프레임을 파일로 저장하기

피클(pickle)은 pandas 전용 데이터 파일로 데이터를 바이너리(2진) 형태로 직렬화한 객체로 저장하는 방법으로 데이터를 오래 보관할 수 있다는 뜻으로 피클이라고 이름이 붙여졌고 피클로 저장하면 스프레드시트 형태보다 더 작은 용량으로 데이터를 저장할 수 있다.

시리즈를 피클 파일로 저장하기

to_pickle() 메소드로 시리즈나 데이터프레임을 피클 파일로 저장할 수 있다.

In [38]:
name = scientists_df.Name
print(type(name))
print(name)
name.to_pickle('./output/scientists_df_name_seried.pickle')

<class 'pandas.Series'>
0       Rosaline Franklin
1          William Gosset
2    Florence Nightingale
3             Marie Curie
4           Rachel Carson
5               John Snow
6             Alan Turing
7            Johann Gauss
Name: Name, dtype: str


데이터프레임을 피클 파일로 저장하기

In [39]:
scientists_df.to_pickle('./output/scientists_df.pickle')

피클 파일을 읽어오기

read_pickle() 메소드로 피클 파일을 읽어올 수 있다.

In [40]:
print(pd.read_pickle('./output/scientists_df_name_seried.pickle'))

0       Rosaline Franklin
1          William Gosset
2    Florence Nightingale
3             Marie Curie
4           Rachel Carson
5               John Snow
6             Alan Turing
7            Johann Gauss
Name: Name, dtype: str


In [41]:
pd.read_pickle('./output/scientists_df.pickle')

,Name,Born,Died,Age,Occupation,born_dt,died_dt,age_days
0,Rosaline Franklin,1920-07-25,1958-04-16,37,Chemist,1920-07-25,1958-04-16,13779 days
1,William Gosset,1876-06-13,1937-10-16,61,Statistician,1876-06-13,1937-10-16,22404 days
2,Florence Nightingale,1820-05-12,1910-08-13,90,Nurse,1820-05-12,1910-08-13,32964 days
3,Marie Curie,1867-11-07,1934-07-04,66,Chemist,1867-11-07,1934-07-04,24345 days
4,Rachel Carson,1907-05-27,1964-04-14,56,Biologist,1907-05-27,1964-04-14,20777 days
5,John Snow,1813-03-15,1858-06-16,45,Physician,1813-03-15,1858-06-16,16529 days
6,Alan Turing,1912-06-23,1954-06-07,41,Computer Scientist,1912-06-23,1954-06-07,15324 days
7,Johann Gauss,1777-04-30,1855-02-23,77,Mathematician,1777-04-30,1855-02-23,28422 days


시리즈나 데이터프레임을 텍스트(csv 또는 tsv) 파일로 저장하기

to_csv() 메소드로 시리즈나 데이터프레임을 텍스트 파일로 저장할 수 있다.  
csv 파일로 저장할때는 to_csv() 메소드의 구분자를 지정하는 sep 옵션의 기본값이 ','이므로 구분자를 별도로 지정하지 않아되 되지만 tsv 파일로 저장할때는 구분자를 반드시 '\t'로 지정해야 한다.

In [49]:
name.to_csv('./output/scientists_df_name_seried.csv')
scientists_df.to_csv('./output/scientists_df.tsv', sep='\t')